rust/kernel/file_operations.rs

```diff
@@ -249,6 +301,70 @@ macro_rules! declare_file_operations {
     };
 }
 
+/// Allows the handling of ioctls defined with the `_IO`, `_IOR`, `_IOW`, and `_IOWR` macros.
+///
+/// For each macro, there is a handler function that takes the appropriate types as arguments.
+pub trait IoctlHandler: Sync {
+    /// Handles ioctls defined with the `_IO` macro, that is, with no buffer as argument.
+    fn pure(&self, _file: &File, _cmd: u32, _arg: usize) -> KernelResult<i32> {
+        Err(Error::EINVAL)
+    }
+
+    /// Handles ioctls defined with the `_IOR` macro, that is, with an output buffer provided as
+    /// argument.
+    fn read(&self, _file: &File, _cmd: u32, _writer: &mut UserSlicePtrWriter) -> KernelResult<i32> {
+        Err(Error::EINVAL)
+    }
+
+    /// Handles ioctls defined with the `_IOW` macro, that is, with an input buffer provided as
+    /// argument.
+    fn write(
+        &self,
+        _file: &File,
+        _cmd: u32,
+        _reader: &mut UserSlicePtrReader,
+    ) -> KernelResult<i32> {
+        Err(Error::EINVAL)
+    }
+
+    /// Handles ioctls defined with the `_IOWR` macro, that is, with a buffer for both input and
+    /// output provided as argument.
+    fn read_write(&self, _file: &File, _cmd: u32, _data: UserSlicePtr) -> KernelResult<i32> {
+        Err(Error::EINVAL)
+    }
+}
+
+/// Dispatches the given ioctl to the appropriate handler based on the value of `cmd`. It also
+/// creates a [`UserSlicePtr`], [`UserSlicePtrReader`], or [`UserSlicePtrWriter`] depending on the
+/// direction of the buffer of the command.
+///
+/// It is meant to be used in implementations of [`FileOperations::ioctl`] and
+/// [`FileOperations::compat_ioctl`].
+pub fn dispatch_ioctl<T: IoctlHandler>(
```

Comment:
I think it'd be more erognomic if this was a method on the argument to `ioctl`, would also make the safety invariants more obvious.

Reply:
Alex, sorry, I'm not sure I follow you. Do you mean to declare say `cmd` as a new type with a `dispatch` method, then pass that as an argument to `ioctl` and `compat_ioctl`?

Reply:
I implemented this, I think it is indeed more ergonomic.